File: examples/python/guard_api_client.py
from __future__ import annotations

import mimetypes
import os
from pathlib import Path
from typing import Any

import requests


class GuardApiClient:
    def __init__(
        self,
        base_url: str | None = None,
        api_key: str | None = None,
        timeout: float = 30.0,
    ) -> None:
        self.base_url = (base_url or os.environ.get("GUARD_API_URL") or "").rstrip("/")
        self.api_key = api_key or os.environ.get("GUARD_API_KEY") or ""
        self.timeout = timeout
        if not self.base_url:
            raise ValueError("GUARD_API_URL is required.")
        if not self.api_key:
            raise ValueError("GUARD_API_KEY is required.")

    def moderate_text(self, text: str, metadata: dict[str, Any] | None = None) -> dict[str, Any]:
        return self._post_json("/moderate/text", {"text": text, "metadata": metadata or {}})

    def moderate_image(
        self,
        file_path: str | Path,
        *,
        image_caption: str = "",
        detected_objects: list[str] | None = None,
        ocr_text: str = "",
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        fields = {
            "image_caption": image_caption,
            "detected_objects": ",".join(detected_objects or []),
            "ocr_text": ocr_text,
            **self._metadata_fields(metadata or {}),
        }
        return self._post_file("/moderate/image", "image", file_path, fields)

    def moderate_audio(
        self,
        file_path: str | Path,
        *,
        transcript_hint: str = "",
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        fields = {
            "transcript_hint": transcript_hint,
            **self._metadata_fields(metadata or {}),
        }
        return self._post_file("/moderate/audio", "audio", file_path, fields)

    def moderate_video(
        self,
        file_path: str | Path,
        *,
        transcript_hint: str = "",
        frames: list[dict[str, Any]] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        import json

        fields = {
            "transcript_hint": transcript_hint,
            "frames": json.dumps(frames or []),
            **self._metadata_fields(metadata or {}),
        }
        return self._post_file("/moderate/video", "video", file_path, fields)

    def _post_json(self, route: str, payload: dict[str, Any]) -> dict[str, Any]:
        response = requests.post(
            f"{self.base_url}{route}",
            headers={
                "Content-Type": "application/json",
                "X-API-Key": self.api_key,
            },
            json=payload,
            timeout=self.timeout,
        )
        return self._response_json(response)

    def _post_file(
        self,
        route: str,
        field_name: str,
        file_path: str | Path,
        fields: dict[str, str],
    ) -> dict[str, Any]:
        path = Path(file_path)
        content_type = mimetypes.guess_type(path.name)[0] or "application/octet-stream"
        with path.open("rb") as handle:
            response = requests.post(
                f"{self.base_url}{route}",
                headers={"X-API-Key": self.api_key},
                files={field_name: (path.name, handle, content_type)},
                data=fields,
                timeout=self.timeout,
            )
        return self._response_json(response)

    @staticmethod
    def _metadata_fields(metadata: dict[str, Any]) -> dict[str, str]:
        allowed = {"content_id", "user_id", "language", "channel", "region"}
        return {key: str(value) for key, value in metadata.items() if key in allowed and value is not None}

    @staticmethod
    def _response_json(response: requests.Response) -> dict[str, Any]:
        try:
            payload = response.json()
        except ValueError:
            payload = {}
        if not response.ok:
            raise RuntimeError(payload.get("detail") or f"Guard API returned {response.status_code}")
        return payload


def decision_action(moderation: dict[str, Any]) -> str:
    return moderation.get("decision", {}).get("action", "review")


def should_publish(moderation: dict[str, Any]) -> bool:
    return decision_action(moderation) == "allow"


def should_hold_for_review(moderation: dict[str, Any]) -> bool:
    return decision_action(moderation) == "review"


def should_block(moderation: dict[str, Any]) -> bool:
    return decision_action(moderation) == "block"

File: examples/node/guard-api-client.mjs
import { readFile } from "node:fs/promises";
import path from "node:path";

const DEFAULT_TIMEOUT_MS = 30_000;

const MIME_BY_EXTENSION = {
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".png": "image/png",
  ".webp": "image/webp",
  ".mp3": "audio/mpeg",
  ".m4a": "audio/mp4",
  ".mp4": "video/mp4",
  ".wav": "audio/wav",
  ".webm": "audio/webm",
  ".mov": "video/quicktime",
};

export class GuardApiClient {
  constructor({
    baseUrl = process.env.GUARD_API_URL,
    apiKey = process.env.GUARD_API_KEY,
    timeoutMs = DEFAULT_TIMEOUT_MS,
  } = {}) {
    if (!baseUrl) {
      throw new Error("GUARD_API_URL is required.");
    }
    if (!apiKey) {
      throw new Error("GUARD_API_KEY is required.");
    }
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.timeoutMs = timeoutMs;
  }

  async moderateText({ text, metadata = {} }) {
    return this.#postJson("/moderate/text", { text, metadata });
  }

  async moderateImage({ filePath, imageCaption = "", detectedObjects = [], ocrText = "", metadata = {} }) {
    const form = await this.#fileForm("image", filePath);
    form.set("image_caption", imageCaption);
    form.set("detected_objects", detectedObjects.join(","));
    form.set("ocr_text", ocrText);
    this.#appendMetadata(form, metadata);
    return this.#postForm("/moderate/image", form);
  }

  async moderateAudio({ filePath, transcriptHint = "", metadata = {} }) {
    const form = await this.#fileForm("audio", filePath);
    form.set("transcript_hint", transcriptHint);
    this.#appendMetadata(form, metadata);
    return this.#postForm("/moderate/audio", form);
  }

  async moderateVideo({ filePath, transcriptHint = "", frames = [], metadata = {} }) {
    const form = await this.#fileForm("video", filePath);
    form.set("transcript_hint", transcriptHint);
    form.set("frames", JSON.stringify(frames));
    this.#appendMetadata(form, metadata);
    return this.#postForm("/moderate/video", form);
  }

  async #postJson(route, body) {
    return this.#request(route, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-Key": this.apiKey,
      },
      body: JSON.stringify(body),
    });
  }

  async #postForm(route, form) {
    return this.#request(route, {
      method: "POST",
      headers: {
        "X-API-Key": this.apiKey,
      },
      body: form,
    });
  }

  async #request(route, init) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await fetch(`${this.baseUrl}${route}`, {
        ...init,
        signal: controller.signal,
      });
      const payload = await response.json().catch(() => ({}));
      if (!response.ok) {
        throw new Error(payload.detail || `Guard API returned ${response.status}`);
      }
      return payload;
    } finally {
      clearTimeout(timeout);
    }
  }

  async #fileForm(fieldName, filePath) {
    const absolutePath = path.resolve(filePath);
    const fileName = path.basename(absolutePath);
    const mimeType = MIME_BY_EXTENSION[path.extname(fileName).toLowerCase()] || "application/octet-stream";
    const bytes = await readFile(absolutePath);
    const blob = new Blob([bytes], { type: mimeType });
    const form = new FormData();
    form.set(fieldName, blob, fileName);
    return form;
  }

  #appendMetadata(form, metadata) {
    for (const [key, value] of Object.entries(metadata)) {
      if (value !== undefined && value !== null) {
        form.set(key, String(value));
      }
    }
  }
}

export function actionForDecision(moderation) {
  return moderation?.decision?.action || "review";
}

export function shouldPublish(moderation) {
  return actionForDecision(moderation) === "allow";
}

export function shouldHoldForReview(moderation) {
  return actionForDecision(moderation) === "review";
}

export function shouldBlock(moderation) {
  return actionForDecision(moderation) === "block";
}

